internal/engine/wazevo/backend/isa/amd64/lower_mem.go

```diff
@@ -1,10 +1,159 @@
+		panic("invalid input")
+	}
+	u64 := uint64(int64(offBase) + x.off + y.off)
+	if u64 != 0 && !lower32willSignExtendTo64(u64) {
+		tmpReg := m.c.AllocateVReg(ssa.TypeI64)
+		m.lowerIconst(tmpReg, u64, true)
+		// Blank u64 as it has been already lowered.
+		u64 = 0
+		// We already know that either rx or ry is invalid,
+		// so we overwrite it with the temporary register.
+		if x.r == regalloc.VRegInvalid {
+			x.r = tmpReg
+		} else {
+			y.r = tmpReg
+		}
+	}
+
+	u32 := uint32(u64)
+	switch {
+	// We assume rx, ry are valid iff offx, offy are 0.
+	case x.r != regalloc.VRegInvalid && y.r != regalloc.VRegInvalid:
+		switch {
+		case x.shift != 0 && y.shift != 0:
+			// Cannot absorb two shifted registers, must lower one to a shift instruction.
+			shifted := m.allocateInstr()
+			shifted.asShiftR(shiftROpShiftLeft, newOperandImm32(uint32(x.shift)), x.r, true)
+			m.insert(shifted)
+
+			return newAmodeRegRegShift(u32, x.r, y.r, y.shift)
+		case x.shift != 0 && y.shift == 0:
+			// Swap base and index.
+			x, y = y, x
+			fallthrough
+		default:
+			return newAmodeRegRegShift(u32, x.r, y.r, y.shift)
+		}
+	case x.r == regalloc.VRegInvalid && y.r != regalloc.VRegInvalid:
+		x, y = y, x
+		fallthrough
+	case x.r != regalloc.VRegInvalid && y.r == regalloc.VRegInvalid:
+		if x.shift != 0 {
+			zero := m.c.AllocateVReg(ssa.TypeI64)
+			m.lowerIconst(zero, 0, true)
+			return newAmodeRegRegShift(u32, zero, x.r, x.shift)
+		}
+		return newAmodeImmReg(u32, x.r)
+	default: // Both are invalid: use the offset.
+		tmpReg := m.c.AllocateVReg(ssa.TypeI64)
+		m.lowerIconst(tmpReg, u64, true)
+		return newAmodeImmReg(0, tmpReg)
+	}
+}
+
+func (m *machine) lowerAddend(x *backend.SSAValueDefinition) addend64 {
+	if x.IsFromBlockParam() {
+		return addend64{x.BlkParamVReg, 0, 0}
+	}
+	// Ensure the addend is not referenced in multiple places; we will discard nested Iadds.
+	op := m.c.MatchInstrOneOf(x, addendsMatchOpcodes[:])
+	if op != ssa.OpcodeInvalid && op != ssa.OpcodeIadd {
+		return m.lowerAddendFromInstr(x.Instr)
+	}
+	return addend64{m.getOperand_Reg(x).r, 0, 0}
+}
+
+// lowerAddendFromInstr takes an instruction returns a Vreg and an offset that can be used in an address mode.
+// The Vreg is regalloc.VRegInvalid if the addend cannot be lowered to a register.
+// The offset is 0 if the addend can be lowered to a register.
+func (m *machine) lowerAddendFromInstr(instr *ssa.Instruction) addend64 {
+	instr.MarkLowered()
+	switch op := instr.Opcode(); op {
+	case ssa.OpcodeIconst:
+		u64 := instr.ConstantVal()
+		if instr.Return().Type().Bits() == 32 {
+			return addend64{regalloc.VRegInvalid, int64(int32(u64)), 0} // sign-extend.
+		} else {
+			return addend64{regalloc.VRegInvalid, int64(u64), 0}
+		}
+	case ssa.OpcodeUExtend, ssa.OpcodeSExtend:
+		input := instr.Arg()
+		inputDef := m.c.ValueDefinition(input)
+		switch input.Type().Bits() {
+		case 64:
+			return addend64{m.getOperand_Reg(inputDef).r, 0, 0}
```

Comment:
does this branch happen?

Reply:
ah yes, good point, I wanted to tell. It happens only in this case:

https://github.com/tetratelabs/wazero/blob/4b85945ee971264cb3a1d9276d7d64f1de867898/internal/engine/wazevo/backend/isa/amd64/lower_mem_test.go#L100-L110

I have included it because it is also there on `collectAddends()` on arm64, but TBH it seems very redundant, I am not even sure it lowers from valid wasm...

Reply:
No, I mean that shouldn't happen in the first place right?

Reply:
at the frontend I meant

Reply:
> also there on collectAddends() on arm64

OK that should be deleted then :D

Reply:
> at the frontend I meant

yeah I don't think so

> OK that should be deleted then :D

:D